core/arch/riscv/mm/core_mmu.c

```diff
@@ -0,0 +1,639 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright 2022 NXP
+ */
+
+#include <assert.h>
+#include <bitstring.h>
+#include <config.h>
+#include <kernel/cache_helpers.h>
+#include <kernel/panic.h>
+#include <kernel/spinlock.h>
+#include <kernel/tee_l2cc_mutex.h>
+#include <kernel/tee_misc.h>
+#include <kernel/tlb_helpers.h>
+#include <mm/core_memprot.h>
+#include <mm/core_mmu.h>
+#include <platform_config.h>
+#include <riscv.h>
+#include <stdlib.h>
+#include <string.h>
+#include <trace.h>
+#include <util.h>
+
+#ifndef RV64
+#error implement
+#endif
+
+#ifndef DEBUG_XLAT_TABLE
+#define DEBUG_XLAT_TABLE 0
+#endif
+
+#if DEBUG_XLAT_TABLE
+#define debug_print(...) DMSG_RAW(__VA_ARGS__)
+#else
+#define debug_print(...) ((void)0)
+#endif
+
+bitstr_t g_asid __nex_bss;
+static unsigned int g_asid_spinlock __nex_bss = SPINLOCK_UNLOCK;
+
+typedef struct {
+	unsigned long entry;
+} pte_t;
+
+typedef struct {
+	pte_t entries[RISCV_PTES_PER_PT];
+} pgt_t;
+
+#define RISCV_MMU_PGT_SIZE	(RISCV_PTES_PER_PT * sizeof(pte_t))
+
+static pgt_t base_xlation_table
+	__aligned(RISCV_PGSIZE)
+	__section(".nozi.mmu.base_table");
+
+static pgt_t xlat_tables[RISCV_MMU_MAX_PGTS]
+	__aligned(RISCV_PGSIZE) __section(".nozi.mmu.l2");
+
+static pgt_t ta_tables[CFG_NUM_THREADS]
+	__aligned(RISCV_PGSIZE) __section(".nozi.mmu.l2");
+
+static int user_va_idx __nex_data = -1;
+
+struct mmu_partition {
+	pgt_t *base_table;
+	pgt_t *xlat_tables;
+	pgt_t *ta_tables;
+	unsigned int pgts_used;
+	unsigned int asid;
+};
+
+static struct mmu_partition default_partition __nex_data  = {
+	.base_table = &base_xlation_table,
+	.xlat_tables = xlat_tables,
+	.ta_tables = ta_tables,
+	.pgts_used = 0,
+	.asid = 0
+};
+
+static inline pte_t *core_mmu_table_get_entry(pgt_t *pgt, unsigned int idx)
+{
+	return &pgt->entries[idx & RISCV_MMU_VPN_MASK];
```

Comment:
The `idx` will be out of bounds always, due to how it is computed, discussed in this [comment](https://github.com/OP-TEE/optee_os/pull/5606#issuecomment-1303252606).

Reply:
That still doesn't make any sense, especially since you actually do mask the index in `core_mmu_pgt_idx()` as one may expect. Perhaps it's time to define what this index really is. In my opinion, this should be the index of a PTE in a specific translation table, not a virtual page index (aka `VPN` in the riscv manual).

`core_mmu_get_entry_primitive()` takes an index of a PTE in a specific translation table and I see no reason to change that.

Reply:
Yes and define what a Page Table is, how to iterate though page levels. Because the index depends on the page level. Yes VPN stills the index in a specific page table but at a specific level. Given the VA we want to map, and given MMU mode is Sv39, We have 3 indexes, for each page table level. VA: 0x80100000 is 4K aligned, we map it at level 0, to get it's PA still needs to iterate though all levels and we compute the index in each level -> VPN[2] = 2; VPN[1] = 256; VPN[0] = 0